File: src/data/courses/dsa/graphs/dsu/provinces-dsu/code/python/provinces_dsu.py
# ============================================================
# Provinces DSU
# ============================================================
class Solution:
    def findCircleNum(self, isConnected) -> int:
        n = len(isConnected)
        parent = list(range(n)); rank = [0]*n
        def find(x):
            if parent[x]!=x: parent[x]=find(parent[x])
            return parent[x]
        def union(x,y):
            rx,ry=find(x),find(y)
            if rx==ry: return
            if rank[rx]<rank[ry]: rx,ry=ry,rx
            parent[ry]=rx
            if rank[rx]==rank[ry]: rank[rx]+=1
        for i in range(n):
            for j in range(i+1,n):
                if isConnected[i][j]==1: union(i,j)
        return sum(1 for i in range(n) if find(i)==i)

File: src/data/courses/dsa/graphs/dsu/provinces-dsu/code/java/ProvincesDsu.java
// ============================================================
// Provinces DSU
// ============================================================
class Solution {
    int[] p, r;
    int find(int x){return p[x]==x?x:(p[x]=find(p[x]));}
    void unite(int x,int y){int a=find(x),b=find(y);if(a==b)return;if(r[a]<r[b]){int t=a;a=b;b=t;}p[b]=a;if(r[a]==r[b])r[a]++;}
    public int findCircleNum(int[][] M) {
        int n=M.length; p=new int[n]; r=new int[n]; for(int i=0;i<n;i++) p[i]=i;
        for(int i=0;i<n;i++) for(int j=i+1;j<n;j++) if(M[i][j]==1) unite(i,j);
        int c=0; for(int i=0;i<n;i++) if(find(i)==i) c++; return c;
    }
}
